javatests/arcs/core/analysis/RecipeGraphFixpointIteratorTest.kt

```diff
@@ -0,0 +1,267 @@
+    private val readCnxn = HandleConnectionSpec("r", HandleMode.Read, TypeVariable("r"))
+    private val writeCnxn = HandleConnectionSpec("w", HandleMode.Write, TypeVariable("w"))
+    private val readerSpec = ParticleSpec(
+        "Reader",
+        listOf(readCnxn).associateBy { it.name },
+        "ReaderLocation"
+    )
+    private val writerSpec = ParticleSpec(
+        "Writer",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val anotherWriterSpec = ParticleSpec(
+        "AnotherWriter",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val readerParticle = Recipe.Particle(
+        readerSpec,
+        listOf(Recipe.Particle.HandleConnection(readCnxn, thing))
+    )
+    private val writerParticle = Recipe.Particle(
+        writerSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+    private val anotherWriterParticle = Recipe.Particle(
+        anotherWriterSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+
+    @Test
+    fun straightLineFlow() {
+        // [Writer] -> (thing) -> [Reader]
+        val recipe = Recipe(
+            "StraightLine",
+            listOf(thing).associateBy { it.name },
+            listOf(readerParticle, writerParticle),
+            "arcId"
+        )
+        val graph = RecipeGraph(recipe)
+        val analyzer = TestAnalyzer(setOf("p:${writerParticle.spec.name}"))
+        analyzer.computeFixpoint(graph)
+        assertThat(analyzer.getValue(writerParticle)?.set).isEmpty()
+        assertThat(analyzer.getValue(thing)?.set)
+            .containsExactly("p:Writer", "p:Writer -> h:thing")
+        assertThat(analyzer.getValue(readerParticle)?.set)
+            .containsExactly("p:Writer", "p:Writer -> h:thing", "h:thing", "h:thing -> p:Reader")
+    }
+
+    @Test
+    fun joinsFlow() {
+        // [Writer] ---------> (thing) -> [Reader]
+        // [AnotherWriter] ------^
+        val recipe = Recipe(
+            "Join",
+            listOf(thing).associateBy { it.name },
+            listOf(readerParticle, writerParticle, anotherWriterParticle),
+            "arcId"
+        )
+        val graph = RecipeGraph(recipe)
+        val analyzer = TestAnalyzer(
+            setOf("p:${writerParticle.spec.name}", "p:${anotherWriterParticle.spec.name}")
+        )
+
+        analyzer.computeFixpoint(graph)
+        assertThat(analyzer.getValue(writerParticle)?.set).isEmpty()
+        assertThat(analyzer.getValue(anotherWriterParticle)?.set).isEmpty()
+        assertThat(analyzer.getValue(thing)?.set)
+            .containsExactly(
+                "p:Writer",
+                "p:AnotherWriter",
+                "p:Writer -> h:thing",
+                "p:AnotherWriter -> h:thing"
+            )
+        assertThat(analyzer.getValue(readerParticle)?.set)
+            .containsExactly(
```

Comment:
Given that you merge all the labels together, it is impossible to have different semantics for: `check p:Writer and check p:AnotherWriter` vs. `check p:Writer or check p:AnotherWriter`.

I believe the TS algorithm tracks every path separately.

Concretely for our needs imagine the following graph:
```
      ----> [ Redactor ] ---> ( redacted ) ---> [Other] ---> (something) ----> [Egress]
    |                                               ^
( needs redacting ) -------------------------------
```

You want to assert that all the paths through which values from (needs redacting) followed to [Egress] went via [Redactor]. I don't think you can check that meaningfully if you don't break down labels per path. 

@csilvestrini for input.

Reply:
The `TestAnalyzer` is solely for the purpose of the testing the fix point iterator. The actual checker will be more precise.

Reply:
Good to know :) Sorry then for deliberating on what's beside the point of this PR, but I wanted to be sure you're aware of that.

Reply:
Getting consensus and having a second pair of eyes on the code is whole point of reviews anyway. :) 

BTW, we don't have to track every path explicitly (which can lead to a blow up) as in the TS algorithm to support the checks we want. We can achieve the same by designing an appropriate lattice.